### PR TITLE
 Generic dataset module and specific s3_datasets module - part 5 (Move DatasetServiceInterface to datasets_base, add property, create first list API for datasets_base) 

### PR DESCRIPTION
**Pull request by dlpzx**
### Feature or Bugfix
- Feature
- Refactoring

### Detail
As explained in the design for #1123 we are trying to implement a generic `datasets_base` module that can be used by any type of datasets in a generic way. 

In this PR we:
- Move DatasetServiceInterface to datasets_base. This interface is used by datasets_sharing to "inject" logic in s3_datasets
- add property dataset_type to the DatasetServiceInterface interface to distinguish which type of dataset this interface applies to.
- create first list API for datasets_base. 👀 This is the most important part. When having multiple types of datasets users will still list all datasets together in several places in the UI (e.g. in listDatasets in DatasetList view, in listDatasetsEnvironment in Environment view) This API calls are not specific to s3_datasets, but generic to any type of dataset. Thus, they should be part of datasets_base. This PR introduces the datasets_list_service, datasetListRepository and includes only one example of API that moves to dataset_base. In next PRs we will move the rest of APIs

### Relates
- #1123 
- #955 

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)?
  - Is the input sanitized?
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization?
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features?
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored?
- Are you introducing any new policies/roles/users?
  - Have you used the least-privilege principle? How?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by dlpzx** [comment]
Testing locally:
- [X] list Datasets view shows list of Datasets - for owned datasets
- [X] list Datasets view shows list of Datasets - for dataset stewards
- [X] list Datasets view shows list of Datasets - for shared datasets

File: backend/dataall/modules/datasets_base/services/dataset_list_service.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+class DatasetListService:
+    _interfaces: List[DatasetServiceInterface] = []
+
+    @classmethod
+    def register(cls, interface: DatasetServiceInterface):
+        cls._interfaces.append(interface)
+
+    @classmethod
+    def _list_all_user_interface_datasets(cls, session, username, groups) -> List:
```
Comment (petrkalos) [inline backend/dataall/modules/datasets_base/services/dataset_list_service.py:18]: Does this return `List[Query]`?

File: backend/dataall/modules/datasets_base/db/dataset_repositories.py [hunk 2]
```diff
@@ -32,3 +35,40 @@ def update_dataset_activity(session, dataset: DatasetBase, username):
...
+    @staticmethod
+    def paginated_all_user_datasets(session, username, groups, all_subqueries, data=None) -> dict:
+        return paginate(
+            query=DatasetListRepository._query_all_user_datasets(session, username, groups, all_subqueries, data),
+            page=data.get('page', 1),
+            page_size=data.get('pageSize', 10),
+        ).to_dict()
+
+    @staticmethod
+    def _query_all_user_datasets(session, username, groups, all_subqueries, filter) -> Query:
```
Comment (petrkalos) [inline backend/dataall/modules/datasets_base/db/dataset_repositories.py:52]: I assume `all_subqueries` is of type `List[Query]`?

**Review by petrkalos — APPROVED** [review]
approving but I'd appreciate if you could add some type hints on the non-trivial arguments

**Pull request merged**